Sprint 10

YanikMax opened the pull request.

Commit:
Свёрстан Launch Screen по макету;
Свёрстан главный экран с лентой изображений.

Commit:
Добавлена градиентная подложка под дату.
Экран профиля сверстан с помощью кода.
Добавлен экран отдельного изображения с возможность шеринга картинки.

Commit:
Исправил констрейнты аватара.
Изменил цвет статус-бара на белый.
Полностью убран код связанный с градиентной подложкой.
Для ImageListCell @IBOutlet сделал приватными.
При импорте UIKit убрал Foundation.
Константу ShowSingleImageSegueIdentifier переименовал (не знаю что на меня нашло).
Извелечение опционала в ImagesListViewController реализовал через guard let.
При верстке кодом использовал предложенное флоу.

Commit:
Для входа в приложение пользователь авторизуется через OAuth2.0;
В проекте создан файл Constants.swift, в нём есть константы AccessKey, SecretKey, RedirectURI и AccessScope;
Свёрстан контроллер WebView;
Реализован KVO — приложение показывает, как обновляется прогресс загрузки веб-страницы;
Проверка необходимости авторизации выделена в отдельный контроллер SplashViewController;
Добавлена структура для декодинга JSON-ответа от POST-запроса авторизации Unsplash — OAuthTokenResponseBody;
Создан новый класс OAuth2Service.

helecta commented on Image Feed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,33 @@
+}
+
+final class AuthViewController: UIViewController {
+    private let ShowWebViewSegueIdentifier = "ShowWebView"
```
Даже такие константы стоит называть со строчной буквы. Да, в учебнике было так, но все-таки правильно называть со строчной. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta replied:
У тебя нет Segue с таким Identifier в сториборде. Поэтому у тебя не во  webViewViewController не проставляется делегат и из-за этого не работает кнопка назад и флоу после авторизации.

helecta commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,101 @@
+import UIKit
+import WebKit
+
+fileprivate let UnsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Константы стоит называть со строчной буквы.

helecta commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,101 @@
+    @IBOutlet private var webView: WKWebView!
+    @IBOutlet private var progressView: UIProgressView!
+    
+    weak var delegate: WebViewViewControllerDelegate?
```
Стоит организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

helecta commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,101 @@
+        
+        webView.navigationDelegate = self
+        
+        var urlComponents = URLComponents(string: UnsplashAuthorizeURLString)!
```
Лучше переписать без force unwrapping, например с guard:
```
guard let urlComponents = URLComponents(string: UnsplashAuthorizeURLString) else { return }
```
Стоит взять себе за привычку если можно без force unwrapping написать, то нужно писать без.

helecta commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,101 @@
+            URLQueryItem(name: "response_type", value: "code"),
+            URLQueryItem(name: "scope", value: AccessScope)
+        ]
+        let url = urlComponents.url!
```
и тут тоже нужно убрать force unwrapping

helecta commented on Image Feed/Constants.swift:
```diff
@@ -0,0 +1,7 @@
+import Foundation
+
+let AccessKey = "@CrGnzEgtPBgkxTHXzgkRmFYVbe4e7pOKsxyfiUAzMHM"
```
Такие константы стоит называть со строчной буквы.  Да, в учебнике было так, но все-таки правильно называть со строчной. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on Image Feed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,96 @@
+        let formatter = DateFormatter()
+        formatter.dateStyle = .long
+        formatter.timeStyle = .none
+        return formatter
```
Если так настроить дата форматтер:
```
private lazy var dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd MMMM yyyy"
        formatter.locale = Locale(identifier: "ru_RU")
        return formatter
    }()
```
То будет выглядеть как в макете, без  `г.`

helecta commented on Image Feed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,89 @@
+    
+    private let descriptionLabel: UILabel = {
+        let label = UILabel()
+        label.text = "Hello, world!"
```
Это поле может быть длинным, в этом случае у тебя текст будет обрезаться.
Лучше тут сделать многострочный вариант строки. Для этого нужно сделать так:
```
label.numberOfLines = 0
```
В этом случае еще нужно добавить trailingAnchor для descriptionLabel

helecta commented on Image Feed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,89 @@
+            avatarImageView.widthAnchor.constraint(equalToConstant: 70),
+            avatarImageView.heightAnchor.constraint(equalToConstant: 70),
+            
+            nameLabel.leadingAnchor.constraint(equalTo: avatarImageView.leadingAnchor),
```
Тут тоже лучше добавить trailingAnchor, чтобы если имя будет слишком длинным, то оно нормально обрезалось

helecta commented on Image Feed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,89 @@
+            logoutButton.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -24),
+            logoutButton.centerYAnchor.constraint(equalTo: avatarImageView.centerYAnchor),
+            
+            loginNameLabel.leadingAnchor.constraint(equalTo: nameLabel.leadingAnchor),
```
И тут тоже лучше добавить trailingAnchor

helecta commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,71 @@
+import UIKit
+
+final class SplashViewController: UIViewController {
+    private let ShowAuthenticationScreenSegueIdentifier = "ShowAuthenticationScreen"
```
константы стоит называть со строчной буквы

helecta commented on Image Feed/ServiceAccessLayer/OAuth2Service.swift:
```diff
@@ -0,0 +1,164 @@
+    
+    private func authTokenRequest(code: String) -> URLRequest {
+        let parameters: [String: Any] = [
+            "client_id": "YOUR_CLIENT_ID",
```
Тут должны быть константы из твоего файла Constants
YOUR_CLIENT_ID = AccessKey
YOUR_CLIENT_SECRET = SecretKey
YOUR_REDIRECT_URI = RedirectURI

helecta commented on Image Feed/Constants.swift:
```diff
@@ -0,0 +1,7 @@
+import Foundation
+
+let AccessKey = "@CrGnzEgtPBgkxTHXzgkRmFYVbe4e7pOKsxyfiUAzMHM"
+let SecretKey = "JjbQCZd3J1MarTVosM_kTtKph3jJAdAZ5RCol8l4e0k"
```
Проверь пожалуйста, что ты правильно скопировал все ключи с сайта `Unsplash`, после того как зарегистрировал приложение на сайте 
Потому что сейчас с этими ключами у тебя не работает авторизация

Commit:
Переименовал константы, которые были написаны с заглавной буквы;
Исправил в сториборде работу кнопки назад;
В WebViewViewController организовал код согласно общепринятым нормам;
Убрал force unwrapping и заменил на guard
Добавил trailingAnchor, если будет слишком длинное имя;
Добавил константы из файла Constants;
Исправил ключ AccessKey (в начале был лишний символ @)
Приложение было зарегистрировано, но слетели галочки разрешений (исправил).
